awslabs/aws-servicebroker
# Bind/Unbind via lambda function

## Overview

This PR adds the ability for the AWS Service Broker to call a lambda function during bind and unbind requests.  

To achieve this, an additional key has been added to the service metadata ("BindViaLambda") and when this has the value `true`, the service broker will also look for a cloud formation Output called `BindLambda` which should be the name or Arn of the Lambda function to use.

During bind the lambda function will receive the Outputs from the CloudFormation template which are usually used as binding credentials, along with the Instance ID and Binding ID and a RequestType (either "bind" or "unbind").  The return value of the lambda function will be presented as the binding credentials to the client.  This, for example, allows us to generate a new binding for each new bind request, with unique credentials.   The included example does this for S3, creating a new IAM User and Access Key and binding it to a policy defined in the CloudFormation template.  You could also this to create, for example, RDS database users.   It is necessary though, that whatever is created is identifiable using the Instance ID and Binding ID, as this is the only data available for the unbind. 

During the unbind the lambda function is responsible for destroying any resources it created during the bind.

## Related Issues

Fixes #1 

## Testing

I have added unit testing as much as convenient to do so, I have also tested this "in anger" on a CloudFoundry based deployment. 

### Notes

I have included a new example YAML file that show's how to dynamically create IAM accounts and Access Keys per binding for an S3 bucket.  

## Testing Instructions

Deploy the service broker, configured to look at a custom S3 bucket.  Place a template there that exercises the new functionality (as per the example file, or one of your own devising).  Provision that template from the service broker, you should see that the lambda function runs and the binding received by the client application contains the binding credentials created by the your lambda function.   Perform an unbind and see that your lambda function is called with the unbind request type.  As the exact behaviour of the lambda function is determined by the template, it's hard to define a test beyond that. 
## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
